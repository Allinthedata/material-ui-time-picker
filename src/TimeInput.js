import React from 'react'
import PropTypes from 'prop-types'
import Dialog, { DialogActions } from 'material-ui/Dialog'
import Button from 'material-ui/Button'
import Input from 'material-ui/Input'
import { withStyles } from 'material-ui/styles'
import TimePicker from './TimePicker'
import { formatHours, twoDigits } from './util'

const styles = {
  header: {
    borderTopLeftRadius: 2,
    borderTopRightRadius: 2
  },
  body: {
    paddingBottom: 20
  }
}

class TimeInput extends React.Component {
  constructor (props) {
    super(props)
    const defaultValue = new Date()
    defaultValue.setSeconds(0)
    defaultValue.setMilliseconds(0)
    this.state = { open: false, value: props.value || props.defaultValue || defaultValue }
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: nextProps.value })
    }
  }

  showDialog = () => this.setState({ open: true, newValue: this.state.value })

  handleChange = (newValue) => {
    this.setState({ newValue })
  }

  handleOk = () => {
    if (this.props.onChange != null) {
      this.props.onChange(this.state.newValue)
    }
    this.setState({ open: false, value: this.state.newValue, newValue: null })
  }

  handleCancel = () => this.setState({ open: false, newValue: null })

  render () {
    const {
      autoOk,
      cancelLabel,
      classes,
      defaultValue, // eslint-disable-line
      disabledProp,
      mode,
      okLabel,
      onChange, // eslint-disable-line
      value: valueProp, // eslint-disable-line
      ...other
    } = this.props

    const { value, newValue } = this.state

    const { hours, isPm } = formatHours(value.getHours(), mode)
    const formattedValue = mode === '12h'
      ? `${hours}:${twoDigits(value.getMinutes())} ${isPm ? 'pm' : 'am'}`
      : `${twoDigits(value.getHours())}:${twoDigits(value.getMinutes())}`

    const { muiFormControl } = this.context
    const disabled = disabledProp || (muiFormControl != null && muiFormControl.disabled)

    return [
      <Input
        {...other}
        disabled={disabled}
        onClick={!disabled ? this.showDialog : null}
        value={formattedValue}
        readOnly
        key='TimeInput-input'
      />,
      <Dialog
        maxWidth='xs'
        open={this.state.open}
        key='TimeInput-dialog'
        onRequestClose={this.handleCancel}
      >
        <TimePicker
          mode={mode}
          value={newValue}
          onChange={this.handleChange}
          onMinutesSelected={autoOk ? this.handleOk : null}
          classes={{ header: classes.header, body: classes.body }}
        />
        <DialogActions>
          <Button onClick={this.handleCancel} color='primary'>{cancelLabel}</Button>
          <Button onClick={this.handleOk} color='primary'>{okLabel}</Button>
        </DialogActions>
      </Dialog>
    ]
  }
}

TimeInput.propTypes = {
  autoOk: PropTypes.bool,
  cancelLabel: PropTypes.string,
  defaultValue: PropTypes.instanceOf(Date),
  mode: PropTypes.oneOf(['12h', '24h']),
  okLabel: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.instanceOf(Date)
}

TimeInput.defaultProps = {
  autoOk: false,
  cancelLabel: 'Cancel',
  mode: '12h',
  okLabel: 'Ok'
}

TimeInput.contextTypes = {
  muiFormControl: PropTypes.object
}

export default withStyles(styles)(TimeInput)
