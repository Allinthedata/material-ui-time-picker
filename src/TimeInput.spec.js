/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import MockDate from 'mockdate'
import TimeInput from './TimeInput'

describe.only('<TimeInput />', () => {
  describe('24h', () => {
    it('matches the snapshot', () => {
      const tree = mount(<TimeInput defaultValue={new Date('2017-10-15T13:37Z')} mode='24h' />)
      expect(tree).toMatchSnapshot()
    })

    it('displays the formatted time', () => {
      const tree = mount(<TimeInput value={new Date(2017, 15, 14, 13, 37, 0, 0)} mode='24h' />)
      expect(getValue(tree)).toBe('13:37')

      const tree2 = mount(<TimeInput value={new Date(2017, 15, 14, 1, 23, 0, 0)} mode='24h' />)
      expect(getValue(tree2)).toBe('01:23')
    })
  })

  describe('12h', () => {
    it('matches the snapshot', () => {
      const tree = mount(<TimeInput defaultValue={new Date('2017-10-15T13:37Z')} mode='12h' />)
      expect(tree).toMatchSnapshot()
    })

    it('displays the formatted time', () => {
      const tree = mount(<TimeInput value={new Date(2017, 15, 14, 13, 37, 0, 0)} mode='12h' />)
      expect(getValue(tree)).toBe('1:37 pm')

      const tree2 = mount(<TimeInput value={new Date(2017, 15, 14, 1, 23, 0, 0)} mode='12h' />)
      expect(getValue(tree2)).toBe('1:23 am')

      const tree3 = mount(<TimeInput value={new Date(2017, 15, 14, 0, 0, 0, 0)} mode='12h' />)
      expect(getValue(tree3)).toBe('12:00 am')

      const tree4 = mount(<TimeInput value={new Date(2017, 15, 14, 0, 0, 0, 0)} mode='12h' />)
      expect(getValue(tree4)).toBe('12:00 am')
    })
  })

  it('supports controlled mode', () => {
    const tree = mount(<TimeInput value={new Date(2017, 15, 14, 13, 37, 0, 0)} mode='24h' />)
    expect(getValue(tree)).toBe('13:37')

    tree.setProps({ value: new Date(2017, 15, 14, 14, 42, 0, 0) })
    expect(getValue(tree)).toBe('14:42')
  })

  describe('uncontrolled mode', () => {
    afterEach(() => {
      MockDate.reset()
    })

    it('supports uncontrolled mode with a default value', () => {
      const tree = mount(<TimeInput defaultValue={new Date(2017, 15, 14, 13, 37, 0, 0)} mode='24h' />)
      expect(getValue(tree)).toBe('13:37')
    })

    it('uses the current time if no value or default value is set', () => {
      MockDate.set(new Date(2017, 15, 14, 13, 37, 0, 0))
      const tree = mount(<TimeInput mode='24h' />)
      expect(getValue(tree)).toBe('13:37')
    })
  })

  // TODO enzyme doesn't support portals yet, add more tests one they work
  // https://github.com/airbnb/enzyme/issues/1150
  //
  // describe('TimePicker dialog', () => {
  //   it('closes when clicking ok', () => {
  //     const tree = mount(<TimeInput />)
  //     tree.simulate('click')
  //     tree.findWhere((e) => e.getDOMNode() != null && e.text() === 'Ok').simulate('click')
  //   })
  //
  //   it('closes when clicking cancel', () => {
  //     const tree = mount(<TimeInput />)
  //     tree.findWhere((e) => e.getDOMNode() != null && e.text() === 'Cancel').simulate('click')
  //   })
  // })
})

function getValue (timeInput) {
  return timeInput.find('input[type="text"]').getDOMNode().value
}
