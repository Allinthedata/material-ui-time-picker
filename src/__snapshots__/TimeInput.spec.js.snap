// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeInput /> 12h matches the snapshot 1`] = `
<withStyles(TimeInput)
  defaultValue={2018-04-14T11:37:00.000Z}
  mode="12h"
>
  <TimeInput
    autoOk={false}
    cancelLabel="Cancel"
    classes={
      Object {
        "body": "TimeInput-body-2",
        "header": "TimeInput-header-1",
      }
    }
    defaultValue={2018-04-14T11:37:00.000Z}
    mode="12h"
    okLabel="Ok"
  >
    <withStyles(Input)
      disabled={false}
      key="TimeInput-input"
      onClick={[Function]}
      readOnly={true}
      value="1:37 pm"
    >
      <Input
        classes={
          Object {
            ".MuiInput-error-6:after": "MuiInput-error-6:after",
            ".MuiInput-inkbar-5.MuiInput-focused-10:after": "MuiInput-inkbar-5.MuiInput-focused-10:after",
            ".MuiInput-inkbar-5:after": "MuiInput-inkbar-5:after",
            ".MuiInput-input-7:-ms-input-placeholder": "MuiInput-input-7:-ms-input-placeholder",
            ".MuiInput-input-7::-moz-placeholder": "MuiInput-input-7::-moz-placeholder",
            ".MuiInput-input-7::-ms-input-placeholder": "MuiInput-input-7::-ms-input-placeholder",
            ".MuiInput-input-7::-webkit-input-placeholder": "MuiInput-input-7::-webkit-input-placeholder",
            ".MuiInput-input-7::-webkit-search-decoration": "MuiInput-input-7::-webkit-search-decoration",
            ".MuiInput-input-7:focus": "MuiInput-input-7:focus",
            ".MuiInput-input-7:invalid": "MuiInput-input-7:invalid",
            ".MuiInput-underline-11.MuiInput-disabled-9:before": "MuiInput-underline-11.MuiInput-disabled-9:before",
            ".MuiInput-underline-11:before": "MuiInput-underline-11:before",
            ".MuiInput-underline-11:hover:not(.MuiInput-disabled-9):before": "MuiInput-underline-11:hover:not(.MuiInput-disabled-9):before",
            "disabled": "MuiInput-disabled-9",
            "error": "MuiInput-error-6",
            "focused": "MuiInput-focused-10",
            "formControl": "MuiInput-formControl-4",
            "fullWidth": "MuiInput-fullWidth-17",
            "inkbar": "MuiInput-inkbar-5",
            "input": "MuiInput-input-7",
            "inputDense": "MuiInput-inputDense-8",
            "inputDisabled": "MuiInput-inputDisabled-13",
            "inputMultiline": "MuiInput-inputMultiline-16",
            "inputSearch": "MuiInput-inputSearch-15",
            "inputSingleline": "MuiInput-inputSingleline-14",
            "label + .MuiInput-formControl-4": "abel + .MuiInput-formControl-4",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-moz-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-moz-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-webkit-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-webkit-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus:-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus:-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-moz-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-moz-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-webkit-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-webkit-input-placeholder",
            "multiline": "MuiInput-multiline-12",
            "root": "MuiInput-root-3",
            "underline": "MuiInput-underline-11",
          }
        }
        disableUnderline={false}
        disabled={false}
        fullWidth={false}
        multiline={false}
        onClick={[Function]}
        readOnly={true}
        type="text"
        value="1:37 pm"
      >
        <div
          className="MuiInput-root-3 MuiInput-inkbar-5 MuiInput-underline-11"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-7 MuiInput-inputSingleline-14"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={true}
            required={undefined}
            rows={undefined}
            type="text"
            value="1:37 pm"
          />
        </div>
      </Input>
    </withStyles(Input)>
    <withStyles(Dialog)
      key="TimeInput-dialog"
      maxWidth="xs"
      onRequestClose={[Function]}
      open={false}
    >
      <Dialog
        classes={
          Object {
            ".MuiDialog-paper-19:focus": "MuiDialog-paper-19:focus",
            "fullScreen": "MuiDialog-fullScreen-24",
            "fullWidth": "MuiDialog-fullWidth-23",
            "paper": "MuiDialog-paper-19",
            "paperWidthMd": "MuiDialog-paperWidthMd-22",
            "paperWidthSm": "MuiDialog-paperWidthSm-21",
            "paperWidthXs": "MuiDialog-paperWidthXs-20",
            "root": "MuiDialog-root-18",
          }
        }
        fullScreen={false}
        fullWidth={false}
        ignoreBackdropClick={false}
        ignoreEscapeKeyUp={false}
        maxWidth="xs"
        onRequestClose={[Function]}
        open={false}
        transition={[Function]}
        transitionDuration={
          Object {
            "enter": 225,
            "exit": 195,
          }
        }
      >
        <withStyles(Modal)
          backdropTransitionDuration={
            Object {
              "enter": 225,
              "exit": 195,
            }
          }
          className="MuiDialog-root-18"
          ignoreBackdropClick={false}
          ignoreEscapeKeyUp={false}
          onBackdropClick={undefined}
          onEscapeKeyUp={undefined}
          onRequestClose={[Function]}
          show={false}
        >
          <Modal
            backdropComponent={[Function]}
            backdropInvisible={false}
            backdropTransitionDuration={
              Object {
                "enter": 225,
                "exit": 195,
              }
            }
            className="MuiDialog-root-18"
            classes={
              Object {
                "hidden": "MuiModal-hidden-26",
                "root": "MuiModal-root-25",
              }
            }
            disableBackdrop={false}
            ignoreBackdropClick={false}
            ignoreEscapeKeyUp={false}
            keepMounted={false}
            modalManager={
              Object {
                "add": [Function],
                "isTopModal": [Function],
                "remove": [Function],
              }
            }
            onBackdropClick={undefined}
            onEscapeKeyUp={undefined}
            onRequestClose={[Function]}
            show={false}
          />
        </withStyles(Modal)>
      </Dialog>
    </withStyles(Dialog)>
  </TimeInput>
</withStyles(TimeInput)>
`;

exports[`<TimeInput /> 24h matches the snapshot 1`] = `
<withStyles(TimeInput)
  defaultValue={2018-04-14T11:37:00.000Z}
  mode="24h"
>
  <TimeInput
    autoOk={false}
    cancelLabel="Cancel"
    classes={
      Object {
        "body": "TimeInput-body-2",
        "header": "TimeInput-header-1",
      }
    }
    defaultValue={2018-04-14T11:37:00.000Z}
    mode="24h"
    okLabel="Ok"
  >
    <withStyles(Input)
      disabled={false}
      key="TimeInput-input"
      onClick={[Function]}
      readOnly={true}
      value="13:37"
    >
      <Input
        classes={
          Object {
            ".MuiInput-error-6:after": "MuiInput-error-6:after",
            ".MuiInput-inkbar-5.MuiInput-focused-10:after": "MuiInput-inkbar-5.MuiInput-focused-10:after",
            ".MuiInput-inkbar-5:after": "MuiInput-inkbar-5:after",
            ".MuiInput-input-7:-ms-input-placeholder": "MuiInput-input-7:-ms-input-placeholder",
            ".MuiInput-input-7::-moz-placeholder": "MuiInput-input-7::-moz-placeholder",
            ".MuiInput-input-7::-ms-input-placeholder": "MuiInput-input-7::-ms-input-placeholder",
            ".MuiInput-input-7::-webkit-input-placeholder": "MuiInput-input-7::-webkit-input-placeholder",
            ".MuiInput-input-7::-webkit-search-decoration": "MuiInput-input-7::-webkit-search-decoration",
            ".MuiInput-input-7:focus": "MuiInput-input-7:focus",
            ".MuiInput-input-7:invalid": "MuiInput-input-7:invalid",
            ".MuiInput-underline-11.MuiInput-disabled-9:before": "MuiInput-underline-11.MuiInput-disabled-9:before",
            ".MuiInput-underline-11:before": "MuiInput-underline-11:before",
            ".MuiInput-underline-11:hover:not(.MuiInput-disabled-9):before": "MuiInput-underline-11:hover:not(.MuiInput-disabled-9):before",
            "disabled": "MuiInput-disabled-9",
            "error": "MuiInput-error-6",
            "focused": "MuiInput-focused-10",
            "formControl": "MuiInput-formControl-4",
            "fullWidth": "MuiInput-fullWidth-17",
            "inkbar": "MuiInput-inkbar-5",
            "input": "MuiInput-input-7",
            "inputDense": "MuiInput-inputDense-8",
            "inputDisabled": "MuiInput-inputDisabled-13",
            "inputMultiline": "MuiInput-inputMultiline-16",
            "inputSearch": "MuiInput-inputSearch-15",
            "inputSingleline": "MuiInput-inputSingleline-14",
            "label + .MuiInput-formControl-4": "abel + .MuiInput-formControl-4",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-moz-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-moz-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-webkit-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7::-webkit-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus:-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus:-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-moz-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-moz-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-ms-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-ms-input-placeholder",
            "label[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-webkit-input-placeholder": "abel[data-shrink=false] + .MuiInput-formControl-4 .MuiInput-input-7:focus::-webkit-input-placeholder",
            "multiline": "MuiInput-multiline-12",
            "root": "MuiInput-root-3",
            "underline": "MuiInput-underline-11",
          }
        }
        disableUnderline={false}
        disabled={false}
        fullWidth={false}
        multiline={false}
        onClick={[Function]}
        readOnly={true}
        type="text"
        value="13:37"
      >
        <div
          className="MuiInput-root-3 MuiInput-inkbar-5 MuiInput-underline-11"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-7 MuiInput-inputSingleline-14"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onChange={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={true}
            required={undefined}
            rows={undefined}
            type="text"
            value="13:37"
          />
        </div>
      </Input>
    </withStyles(Input)>
    <withStyles(Dialog)
      key="TimeInput-dialog"
      maxWidth="xs"
      onRequestClose={[Function]}
      open={false}
    >
      <Dialog
        classes={
          Object {
            ".MuiDialog-paper-19:focus": "MuiDialog-paper-19:focus",
            "fullScreen": "MuiDialog-fullScreen-24",
            "fullWidth": "MuiDialog-fullWidth-23",
            "paper": "MuiDialog-paper-19",
            "paperWidthMd": "MuiDialog-paperWidthMd-22",
            "paperWidthSm": "MuiDialog-paperWidthSm-21",
            "paperWidthXs": "MuiDialog-paperWidthXs-20",
            "root": "MuiDialog-root-18",
          }
        }
        fullScreen={false}
        fullWidth={false}
        ignoreBackdropClick={false}
        ignoreEscapeKeyUp={false}
        maxWidth="xs"
        onRequestClose={[Function]}
        open={false}
        transition={[Function]}
        transitionDuration={
          Object {
            "enter": 225,
            "exit": 195,
          }
        }
      >
        <withStyles(Modal)
          backdropTransitionDuration={
            Object {
              "enter": 225,
              "exit": 195,
            }
          }
          className="MuiDialog-root-18"
          ignoreBackdropClick={false}
          ignoreEscapeKeyUp={false}
          onBackdropClick={undefined}
          onEscapeKeyUp={undefined}
          onRequestClose={[Function]}
          show={false}
        >
          <Modal
            backdropComponent={[Function]}
            backdropInvisible={false}
            backdropTransitionDuration={
              Object {
                "enter": 225,
                "exit": 195,
              }
            }
            className="MuiDialog-root-18"
            classes={
              Object {
                "hidden": "MuiModal-hidden-26",
                "root": "MuiModal-root-25",
              }
            }
            disableBackdrop={false}
            ignoreBackdropClick={false}
            ignoreEscapeKeyUp={false}
            keepMounted={false}
            modalManager={
              Object {
                "add": [Function],
                "isTopModal": [Function],
                "remove": [Function],
              }
            }
            onBackdropClick={undefined}
            onEscapeKeyUp={undefined}
            onRequestClose={[Function]}
            show={false}
          />
        </withStyles(Modal)>
      </Dialog>
    </withStyles(Dialog)>
  </TimeInput>
</withStyles(TimeInput)>
`;
